libc/string/string_lenchr.c

```diff
@@ -96,61 +99,215 @@ TEST(string_len, ascii)
+	char holder[45] = { 0 };
+	int sz;
+
+	/* Checking if both functions recognize the holder as a different set of elements */
+	sz = sizeof(pangram) - 1;
+
+	memcpy(holder, pangram, sz + 1);
+
+	/* Checking if both functions recognize the holder as the same set of elements */
+	TEST_ASSERT_EQUAL_INT(0, strcspn(pangram, holder));
+	TEST_ASSERT_EQUAL_INT(sz, strspn(pangram, holder));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(pangram, &holder[sz / 2]));
+	/* One because strspn found space as*/
+	TEST_ASSERT_EQUAL_INT(1, strspn(pangram, &holder[sz / 2]));
+
+	TEST_ASSERT_EQUAL_INT(sz, strcspn(pangram, &holder[sz]));
+}
+
+
+TEST(string_spn, ascii)
+{
+	char supportCharSet[BUFF_SIZE] = { 0 },
+		 reversCharSet[BUFF_SIZE] = { 0 },
+		 *asciiStr;
+	int i;
+
+	asciiStr = testdata_createCharStr(BUFF_SIZE + 1);
+
+
+	for (i = 1; i < BUFF_SIZE; i++) {
+		supportCharSet[i - 1] = i;
+		reversCharSet[i - 1] = BUFF_SIZE - i;
+
+		/*
+		 * In this case we need to use fully filled set for strcspn
+		 * because it counts size based on elements that are not
+		 * present in himself
+		 */
+		TEST_ASSERT_EQUAL_INT(i - 1, strcspn(&asciiStr[1], &asciiStr[i]));
+		TEST_ASSERT_EQUAL_INT(i, strspn(&asciiStr[1], supportCharSet));
+
+		/* Checking if we reverse order strcspn and strspn will find elements that are correct */
+		TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strcspn(&asciiStr[1], &reversCharSet[i]));
+
+		if (i == 127) {
+			TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strspn(&asciiStr[1], reversCharSet));
+		}
+		else {
+			TEST_ASSERT_EQUAL_INT(0, strspn(&asciiStr[1], reversCharSet));
+		}
+	}
+
+	free((void *)asciiStr);
+}
+
+
+TEST(string_spn, not_ascii)
+{
+	unsigned char charSet[BUFF_SIZE] = { 0 };
 	int sz, i;
 
-	/* The length of string is not clearly restricted, so we test one of bigger value, which may be used  */
-	for (i = 0; i < PATH_MAX - 1; i++) {
-		bigstr[i] = 'A';
+	/* Checking  out of ASCII bytes */
+	for (i = sizeof(charSet); i < sizeof(charSet) * 2 - 1; i++) {
+		charSet[i - sizeof(charSet)] = i;
 	}
-	bigstr[i] = '\0';
+
+	sz = sizeof(charSet) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strcspn((const char *)charSet, ""));
+	TEST_ASSERT_EQUAL_INT(sz, strspn((const char *)charSet, (const char *)charSet));
+}
+
+
+TEST(string_spn, big)
+{
+	char bigstr[PATH_MAX] = { 0 },
+		 support[PATH_MAX] = { 0 };
+	int sz;
+
+	/*
+	 * The length of the string is not restricted,
+	 * so we test one of the bigger values, which may be used
+	 * with remembering the last element must be a null term
+	 * zero to collect the data as the string from
+	 * declared space, that's why we memset one place less than its size
+	 */
 
 	sz = sizeof(bigstr) - 1;
-	TEST_ASSERT_EQUAL_INT(sz, strlen(bigstr));
-	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(bigstr, sz - 1));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz + 1));
+	memset(bigstr, 'a', sz);
+
+	TEST_ASSERT_EQUAL_INT(sz, strcspn(bigstr, ""));
+	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, bigstr));
```

Comment:
If we have buffers provided we can add there a little bit more checks (I know that some of them were placed in empty_output case, but when using big string as s1 it could behave in a different way). Above that the utility of placing more than one same byte in s2 should be tested in the separate case (and probably is), that's why "a" would be sufficient there.

```suggestion
	TEST_ASSERT_EQUAL_INT(sz, strcspn(bigstr, ""));
	TEST_ASSERT_EQUAL_INT(0, strcspn(bigstr, "a"));
	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, "a"));
	TEST_ASSERT_EQUAL_INT(0, strspn(bigstr, ""));
```